staging/src/k8s.io/apiserver/pkg/storage/feature/feature_support_checker.go

```diff
@@ -43,67 +46,98 @@ type FeatureSupportChecker interface {
 	switch feature {
 	case storage.RequestWatchProgress:
 		f.lock.Lock()
 		defer f.lock.Unlock()
 
-		return ptr.Deref(f.progressNotifySupported, false), nil
+		return ptr.Deref(f.progressNotifySupported, false)
 	default:
-		return false, fmt.Errorf("feature %q is not implemented in DefaultFeatureSupportChecker", feature)
+		runtime.HandleError(fmt.Errorf("feature %q is not implemented in DefaultFeatureSupportChecker", feature))
+		return false
 	}
 }
 
 // CheckClient accepts client and calculate the support per endpoint and caches it.
-// It will return at any point if error happens or one endpoint is not supported.
-func (f *defaultFeatureSupportChecker) CheckClient(ctx context.Context, c client, feature storage.Feature) error {
+func (f *defaultFeatureSupportChecker) CheckClient(ctx context.Context, c client, feature storage.Feature) {
 	switch feature {
 	case storage.RequestWatchProgress:
-		return f.clientSupportsRequestWatchProgress(ctx, c)
+		f.checkClient(ctx, c)
 	default:
-		return fmt.Errorf("feature %q is not implemented in DefaultFeatureSupportChecker", feature)
-
+		runtime.HandleError(fmt.Errorf("feature %q is not implemented in DefaultFeatureSupportChecker", feature))
 	}
 }
 
-func (f *defaultFeatureSupportChecker) clientSupportsRequestWatchProgress(ctx context.Context, c client) error {
+func (f *defaultFeatureSupportChecker) checkClient(ctx context.Context, c client) {
+	// start with 10 ms, multiply by 2 each step, until 15 s and stays on 15 seconds.
+	delayFunc := wait.Backoff{
+		Duration: 10 * time.Millisecond,
+		Cap:      15 * time.Second,
+		Factor:   2.0,
+		Steps:    11}.DelayFunc()
 	f.lock.Lock()
 	defer f.lock.Unlock()
-
 	for _, ep := range c.Endpoints() {
-		supported, err := f.supportsProgressNotifyEndpointLocked(ctx, c, ep)
-		if err != nil {
-			return err
-		}
-		if !supported {
-			f.progressNotifySupported = ptr.To(false)
-			return nil
+		if _, found := f.checkingEndpoint[ep]; found {
+			continue
 		}
+		f.checkingEndpoint[ep] = struct{}{}
+		go func(ep string) {
+			err := delayFunc.Until(ctx, true, true, func(ctx context.Context) (done bool, err error) {
+				defer runtime.HandleCrash()
+				internalErr := f.clientSupportsRequestWatchProgress(ctx, c, ep)
+				return internalErr != nil, nil
+			})
+			if err != nil {
+				klog.ErrorS(err, "Failed to check if RequestWatchProgress is supported by etcd after retrying")
+			}
+			supports := f.Supports(storage.RequestWatchProgress)
+			if !supports {
+				klog.Info("RequestWatchProgress feature is not supported by provided etcd. This will prevent ConsistentListFromCache from working")
+			}
+		}(ep)
+	}
+}
+
+func (f *defaultFeatureSupportChecker) clientSupportsRequestWatchProgress(ctx context.Context, c client, ep string) error {
+	f.lock.Lock()
+	defer f.lock.Unlock()
+
+	supported, err := f.supportsProgressNotifyEndpointLocked(ctx, c, ep)
+	if err != nil {
+		return err
+	}
+	if !supported {
```

Comment:
> By checking only the supported variable the state can flap from on to off which is not ideal.

You cannot avoid it without blocking apiserver start on all etcd instances being available.

Reply:
Please take a look at tests we have a few scenario for that.

Reply:
> Also, I think I've asked this question before (but I cannot find it). The doc for the `Supports` says `Returns true if all endpoints in etcd clients are supporting the feature`. I think that the `clientSupportsRequestWatchProgress` method doesn't ensure that.

Can you be more specific?

Reply:
I am concerned about the dynamic nature of the function itself.
The fact that it can be set to `on` for arbitrary time and then to `off`.
It’s also about not really knowing the number of endpoints we are waiting for.
The fact that checking some endpoints may fail (error) and about the fact that I am not able to tell whether the checker is still initialising or if the feature is not supported.

Reply:
This behavior is enough for consistent reads from cache. If there is a change if `supported` the only bad effect would be very short time where APF calculations were incorrect.

Reply:
Today I know that without this feature, in some cases when streaming is used, the server might deadlock. So it would be useful to know whether this feature is supported or not so that streaming could be disabled.

Reply:
Ok, makes sense that streaming will need some different approach, but let's not try to solve all the problems at the same time. When we have consistent lists from cache working we would be happy to help with WatchList.

Reply:
From what I've managed to figure out so far, solving this problem won't be easy and might even require changing the entire approach/implementation :(

Reply:
Let's unblock one of the features and not block both..